### PR TITLE
[pyspark] Cleanup data processing.

### PR DESCRIPTION
**Pull request by trivialfis**
This PR does some cleanups for the data processing precedures for the newly gained pyspark interface.

- Use numpy stack for handling list of arrays.
- Reuse concat function from dask.
- Remove unused code.
- Use iterator for prediction to avoid initializing xgboost model (pickle is not cheap).

To-dos:
- [ ] Test for different types of inputs.

I'm not entirely sure how to work with sparse data since it hasn't been supported yet in current codebase.

@WeichenXu123

### CODE REVIEW

**Comment by trivialfis** [comment]
@wbo4958 Please upstream your work on the support of cuDF when you are available.

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [inline python-package/xgboost/spark/data.py:166]: Is `DeviceQuantileDMatrix` only used in GPU training ?

Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py:166]: Shall we replace CPU side DMatrix with https://github.com/dmlc/xgboost/issues/7890 ?

Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py:166]: Does Quantile dmatrix reduce training result precision ?

**Comment by wbo4958** [comment]
Seems this PR has changed the original way to build DMatrix, can you merge it first, then I will add the whole GPU pipeline support? @trivialfis

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [inline python-package/xgboost/spark/data.py:70]: I think you forget to apply other dmatrix params:
https://github.com/dmlc/xgboost/blob/e28f6f66577eb31451e5d3fbf4be0a21626f7372/python-package/xgboost/spark/core.py#L565

File: python-package/xgboost/spark/data.py
Comment (wbo4958) [inline python-package/xgboost/spark/data.py:9]: May I ask what this function does?

Comment (wbo4958) [reply python-package/xgboost/spark/data.py:9]: So, that is to say,  the ETL converts the single columns into column arrays, and then pandas_udf converts the column of arrays back into a single array? It looks like this operation is redundant and causes the performance issue?

By the way, does that "converts a column of arrays into a single array" cause memory copy?

Comment (trivialfis) [reply python-package/xgboost/spark/data.py:9]: Yup, that's just the convention of spark I guess, a way to handle `feature_col`. We can support `feature_cols` later and recommend it in our document. But still, we need to handle the `feature_col` gracefully.

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [inline python-package/xgboost/spark/data.py:39]: we can use `nonlocal n_features` in append function,
so that `n_features` in outer scope can be a `int` type and inside `append` we can still modify it.

File: python-package/xgboost/spark/data.py
Comment (trivialfis) [reply python-package/xgboost/spark/data.py:166]: https://github.com/dmlc/xgboost/issues/7890 is still working in progress and it's only applicable to hist tree method. I will update it once ready.

> Does Quantile dmatrix reduce training result precision

For CPU no. But for GPU it might generate slight different result due to the fact that GPU uses sliding window for quantile summary generation while the CPU processes a single sample at a time.

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py:9]: If `stack_df` input is `pd.DataFrame`, that means the features column  in train dataset must be a spark struct type. @trivialfis Do we plan to support spark struct type as features input ?

Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py:9]: Seems we should use `array = s.to_numpy()` instead of `array = s.values`. WDYT ?

**Comment by WeichenXu123** [comment]
@trivialfis 
It will be great if you can prioritize this PR,
so that it can unblock follow-up work including data iteration input, external memory, sparse vector support, etc.

**Comment by trivialfis** [comment]
Apologies for the delay. I was trying to get https://github.com/dmlc/xgboost/pull/8050 ready and mitigate the memory usage surge with it.  Will focus on this PR now.

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [inline python-package/xgboost/spark/data.py:16]: I made a simple test:
```
>>> p1 = pd.DataFrame({'a':[[3,4],[5,6],[7,8]]})
>>> p1.a.to_numpy().shape
(3,)
>>> p1.a.to_numpy().reshape(3,1)
array([[list([3, 4])],
       [list([5, 6])],
       [list([7, 8])]], dtype=object)
```
After reshape, each element becomes list, does this hurt performance ?

Comment (WeichenXu123) [inline python-package/xgboost/spark/data.py:16]: Do we need `reshape` here ?

e.g.

```
>>> p1 = pd.DataFrame({'a':[np.array([3,4]), np.array([4,5])]})
>>> p1
        a
0  [3, 4]
1  [4, 5]
>>> p1.a
>>> np.stack(p1.a.to_numpy())
array([[3, 4],
            [4, 5]])
```
I don't use `reshape` and directly stack series values , also got correct result.

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [inline python-package/xgboost/spark/data.py:188]: For DeviceQuantileDMatrix, it supports iterator as input, we don't need to load the whole partition data into memory, right ?

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [inline python-package/xgboost/spark/data.py:156]: This method is very similar with `create_dmatrix_from_partitions`, we can merge them into one function

**Comment by trivialfis** [comment]
@wbo4958 @WeichenXu123 Could you please take another look?

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py]: Current code (for GPU) is:
1. load whole partition data into memory arrays
2. construct an iterator from memory arrays in step (1)
3. construct DMatrix from iterator in step (2)

What I propose it not to write data to disk, but change step (1)(2) into:  construct an iterator from spark python UDF data iterator. This avoid load data into memory in (1) step.

We can do this improvement in follow-up PR.

**Comment by WeichenXu123** [comment]
overall good, @trivialfis Have you test whether it causes performance regression ?

File: python-package/xgboost/spark/data.py
Comment (trivialfis) [reply python-package/xgboost/spark/data.py]: @WeichenXu123 I think the data is still loaded into memory. We need to iterate through the data twice (notice the reset function). Unless the spark iterator can be reset and start again, we need to cache the data somewhere.

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [inline python-package/xgboost/spark/data.py:102]: feature_cols argument is not used when calling create_dmatrix_from_partitions in current code ?

Comment (trivialfis) [reply python-package/xgboost/spark/data.py:102]: No, @wbo4958 will followup on adding that parameter for GPU dataframe.

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [inline python-package/xgboost/spark/data.py:166]: why use `DeviceQuantileDMatrix` only when feature_cols is not None ?

Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py:166]: Use DeviceQuantileDMatrix or not should be controlled by `use_gpu` param if I understand correctly?

Comment (trivialfis) [reply python-package/xgboost/spark/data.py:166]: So far `QuantileDMatrix` is GPU only until https://github.com/dmlc/xgboost/pull/8050 is merged. `feature_cols` will also be a GPU-only thing for the recent future. (Actually I'm not entirely sure about the `feature_cols` parameter that's available on Jvm packages, I don't know how does it work with spark ml pipeline, I will leave these questions to @wbo4958 ).

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [inline python-package/xgboost/spark/data.py:169]: Error...
You forgot to pass the additional `kwargs` for Dmatrix again..
we might need to add test to guard this..

Comment (trivialfis) [reply python-package/xgboost/spark/data.py:169]: !

Comment (trivialfis) [reply python-package/xgboost/spark/data.py:169]: Thank you for the catch! I fixed it and added a test with feature_types.

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py:166]: We don't need to support `feature_cols` param in xgboost pyspark estimator. The transformers in spark pipeline can assemble multiple feature columns into one vector type feature column.

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py]: ah, got it , spark udf data iterator does not support reset. that's bad.

Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py]: Previously I thought DMatrix should accept an one-shot iterator, it exhausts the iterator and cache all data into DMatrix format in memory, and then xgboost training can use the DMatrix data in mutilple passes. Why DMatrix iteration API is not designed this way ?

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py:169]: Though we have tests, I still suggest do this change https://github.com/dmlc/xgboost/pull/8088#discussion_r927455747 because in future if we add more code branches some branch might miss test coverage

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py:102]: I think we don't need this 
https://github.com/dmlc/xgboost/pull/8088#discussion_r927486213

File: python-package/xgboost/spark/data.py
Comment (trivialfis) [reply python-package/xgboost/spark/data.py]: @WeichenXu123 Apologies for the confusion here. Here's a little bit of clarification on the difference between different DMatric constructors:
- Normal DMatrix: `DMatrix(np.array(...))`
  It accepts one single batch of data, and constructs an internal CSR representation that can be used by all algorithms. It has 2 issues with distributed training: firstly, it needs to concatenate all partitions, which doubles the memory usage, secondly, it's a csr, which can triple the memory usage when input is dense.
- `QuantileDMatrix`: This is designed to be as "inplace" as possible and can only be used by GPU hist.  It omits the CSR representation and constructs the histogram index directly from external data. The iterator concept you see there is designed for distributed system that handles data in terms of partitions, the sole purpose of accepting an iterator instead of a whole blob here is to avoid the `concatentate` function on the input data, which doubles the memory. (`concat_or_none` in this PR.)
- External memory: `DMatrix(iterator)`. Yes, using CPU `hist`/`approx` DMatrix caches the batches on disk and iterates through it. But it's quite slow since we need to iterate through the data 3-4 times for each layer of the tree, you can estimate the time usage based on the throughput of your hard disk.

File: python-package/xgboost/spark/data.py
Comment (trivialfis) [reply python-package/xgboost/spark/data.py:166]: I think @wbo4958 wanted to avoid the vector assembler since xgboost needs to undo it. (the `stack_series`). He wanted to have the input data just like other python libraries which 1 column per feature. I think it's a reasonable optimization looking solely on XGBoost, but might not play well with spark ml pipelines.

Comment (trivialfis) [reply python-package/xgboost/spark/data.py:166]: I will remove it for now.

Comment (trivialfis) [reply python-package/xgboost/spark/data.py:166]: Actually, I will keep it for now and leave it as a followup.

File: python-package/xgboost/spark/data.py
Comment (trivialfis) [reply python-package/xgboost/spark/data.py]: Changed according to your suggestion.

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py:166]: > but might not play well with spark ml pipelines.

It does not play well with both spark ml pipelines and other spark meta-estimators such as CrossValidator, OneVsRest, etc..

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py]: Got it.

so QuantileDMatrix this does not support sparse data input ?

Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py]: I would like to assign all data iterator related work to you because you have deeper understanding on this. :)

File: python-package/xgboost/spark/data.py
Comment (trivialfis) [reply python-package/xgboost/spark/data.py:166]: Got it, I will remove it. cc @wbo4958 .

**Review by WeichenXu123 — APPROVED** [review]
LGTM

File: python-package/xgboost/spark/data.py
Comment (trivialfis) [reply python-package/xgboost/spark/data.py]: > so QuantileDMatrix this does not support sparse data input ?

It will suppotr CSR once the CPU impl is merged.

> I would like to assign all data iterator related work to you because you have deeper understanding on this. :)

Got it.

File: python-package/xgboost/spark/data.py
Comment (wbo4958) [reply python-package/xgboost/spark/data.py:166]: @WeichenXu123, yeah, it indeed does not play well with some spark ml pipelines, but we can give users some warning about that. Some users may not use spark ml pipeline or meta-estimators at all, they may just want to train an XGBoost model. I think we should provide this.

**Review by wbo4958 — APPROVED** [review]
LGTM

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py:166]: @wbo4958 
OK. But I doubt "1 column per feature" in spark dataframe and pass many columns to python UDF it might not have good performance when there're many features. We'd better benchmark it.

Comment (wbo4958) [reply python-package/xgboost/spark/data.py:166]: @WeichenXu123, Yes, we will do some benchmark testing. BTW, may I ask why "pass many columns to python UDF it might not have good performance when there're many features", is the penalty happening on the final data size of ArrowRecordBatch?

File: python-package/xgboost/spark/data.py
Comment (wbo4958) [inline python-package/xgboost/spark/data.py:173]: Seems DeviceQuantileDMatrix needs the extra "max_bin", while kwargs does not contain this. But it's ok to file a follow-up for this

Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py:173]: Let's mark a TODO there, to prevent forget it.

**Comment by WeichenXu123** [comment]
If use `feature_cols`, we should check that this only supports GPU (e.g. `use_gpu` param must be True). @wbo4958 You should add this logic in follow-up PR.

**Comment by WeichenXu123** [comment]
@trivialfis Have you test whether it causes performance regression ? The code has big changes.

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py:173]: @wbo4958 @trivialfis Does `DMatrix` need `max_bin` param ? 
And could you help me check whether DMatrix need other param settings, I might miss some params in my PR.

**Comment by trivialfis** [comment]
> Have you test whether it causes performance regression ? The code has big changes.

There's no performance change. Actually, I was hoping numpy can deliver something faster than python list. Seems not much can be done there.

I implemented a very primitive pyspark interface for xgboost for self-education purposes before. I handled the `features` column by expanding it into multiple columns first before passing it to the local fit function. But it handles only vector and array. Wondering if it's a possible future change.

File: python-package/xgboost/spark/data.py
Comment (WeichenXu123) [reply python-package/xgboost/spark/data.py:166]: > is the penalty happening on the final data size of ArrowRecordBatch?

Yes, that's my concern. But I am not expert on this, maybe I am wrong. Let's benchmark.

**Comment by WeichenXu123** [comment]
> I handled the features column by expanding it into multiple columns first before passing it to the local fit function

I guess it might not increase performance, but has risk to downgrade performance

**Comment by wbo4958** [comment]
> If use `feature_cols`, we should check that this only supports GPU (e.g. `use_gpu` param must be True). @wbo4958 You should add this logic in follow-up PR.

Sure, actually, I had the follow up PR up

**Comment by trivialfis** [comment]
>  guess it might not increase performance, but has risk to downgrade performance

I thought spark might handle large number of samples faster than Python. Python list is usually really slow compared to optimized procedures

**Pull request merged**
